Create 2024-05-21-solidity-0.8.26-release-announcement.md (#101)
ethereum/solidity-website: src/posts/2024-05-21-solidity-0.8.26-release-announcement.md

---
title: 'Solidity 0.8.26 Release Announcement'
date: '2024-05-21'
author: Solidity Team
category: Releases
---

We are excited to announce the release of the Solidity Compiler [v0.8.26](https://github.com/ethereum/solidity/releases/tag/v0.8.26).
This newest version of the compiler brings support for custom errors in `require`, improved default Yul Optimizer sequence that will speed up compilation via IR, several bugfixes, and more!

> ## Important Note

Reviewer: What's the important note?

## Notable Features

### Custom errors support in `require`

Custom errors in Solidity provide a convenient and gas-efficient way to explain to the user why an operation failed.
Solidity 0.8.26 introduces a highly anticipated feature that enables the usage of errors with `require` function.

The `require` function in pre 0.8.26 versions provided two overloads:
- `require(bool)` which will revert without any data (not even an error selector).
- `require(bool, string)` which will revert with `Error(string)`.

In this release we are introducing a new overload to support custom errors:
- `require(bool, error)` which will revert with the custom, user supplied error provided as the second argument.

Let's look at an example to understand the usage of the `require` function with custom errors:

> ```solidity=

Reviewer 2: I normally just use `solidity` - not sure weather `solidity=` is really valid..

Reviewer 2: Ok, will fix that too.

// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.26;

/// Insufficient balance for transfer. Needed `required` but only
/// `available` available.
/// @param available balance available.
/// @param required requested amount to transfer.
error InsufficientBalance(uint256 available, uint256 required);

// This will only compile via IR
contract TestToken {
    mapping(address => uint) balance;
    function transferWithRequireError(address to, uint256 amount) public {
        require(
            balance[msg.sender] >= amount,
            InsufficientBalance(balance[msg.sender], amount)
        );
        balance[msg.sender] -= amount;
        balance[to] += amount;
    }
    // ...
}
```

> Note that, just like in the previously available overloads of `require`, arguments are evaluated unconditionally, so take special care to make sure that they are no expressions with unexpected side-effects.

Reviewer 3: ```suggestion
Note that, just like in the previously available overloads of `require`, arguments are evaluated unconditionally, so take special care to make sure that they are not expressions with unexpected side-effects.
```

Our analysis showed, however, that the result of the first pass is almost always very close to the final result and subsequent repetitions of the main loop contribute only a little.
While simple removal of the loop gives results that are still noticeably worse than with the old sequence, with some experimentation we managed to create a [new sequence](https://github.com/ethereum/solidity/blob/v0.8.26/libsolidity/interface/OptimiserSettings.h#L44-L64) that provides comparable optimization quality in a single pass.

For example, this is how bytecode size after each optimization step in the current sequence looks like for some of the sample contracts we analyzed:
![bytecode-size-vs-optimization-time-v0.8.25](https://gist.github.com/assets/137030/e46b4505-e9ae-4b97-90d7-930ef20383a0)

The new sequence stops much earlier:
![bytecode-size-vs-optimization-time-v0.8.26](https://gist.github.com/assets/137030/9c326b84-7d8e-4e5b-8413-0eaa3353b3ef)

Similarly for runtime gas.
Current sequence:
![runtime-gas-vs-optimization-time-v0.8.25](https://gist.github.com/assets/137030/2dfdf7cf-1670-4fb9-b048-9e8f2081dc1d)

New sequence:

> ![runtime-gas-vs-optimization-time-v0.8.26](https://gist.github.com/assets/137030/e3a5b9c6-8441-46e2-989e-b04998d06047)

Reviewer 3: Images are still in my gist, not in the repo.

Reviewer 3: By the way, note that the current plots do not render in github's preview. I do hope they will show up in the post, but it would be safer to have them in the repo even just for that reason.

Reviewer 2: I uploaded the files and updated it. Is it important to have `.svg` ending?

Reviewer 3: Probably better to have it. Any good reason not to give them proper names and extensions?

Also, a subdir in `solidity-website/public/img/` sounds like a better place for them. That's where we store the survey images as well. `src/posts/` is not where images go. Not sure Jekyll will even handle them there.

Finally, you also need absolute links I think. Check how e.g. https://github.com/ethereum/solidity-website/blob/main/src/posts/2024-04-03-solidity-developer-survey-2023-results.md handles images.

Reviewer 3: For the file names, please just use what you see in the square brackets in the snippet above. These were actually the original file names before I uploaded them to the gist.

For this reason, while the new sequence can also be used with older compilers, we recommend extreme care while doing so.
In particular, the new sequence is susceptible to the [FullInliner Non-Expression-Split Argument Evaluation Order Bug](https://soliditylang.org/blog/2023/07/19/full-inliner-non-expression-split-argument-evaluation-order-bug/), which is not an issue for the recent versions, but would cause problems on versions older than v0.8.21.

## Full Changelog

### Language Features

* Introduce a new overload `require(bool, Error)` that allows usage of `require` functions with custom errors. This feature is available in the `via-ir` pipeline only.

### Compiler Features

* SMTChecker: Create balance check verification target for CHC engine.
* Yul IR Code Generation: Cheaper code for reverting with errors of a small static encoding size.
* Yul Optimizer: New, faster default optimizer step sequence.

### Bugfixes

> * Commandline Interface: Fix ICE when the optimizer is disabled and an empty/blank string is used for ``--yul-optimizations` sequence.

Reviewer 3: Is it like that in the changelog? Do we have to correct it there as well?
```suggestion
* Commandline Interface: Fix ICE when the optimizer is disabled and an empty/blank string is used for `--yul-optimizations` sequence.
```

Reviewer 3: Seems [fine in the changelog](https://github.com/ethereum/solidity/blob/v0.8.26/Changelog.md?plain=1#L14) so only need a correction here.

Author: Yeah, this is the default auto quotation in hackmd. So I missed to remove the extra from the autocorrect.

The default sequence in Solidity v0.8.25 included the following steps:
```
dhfoDgvulfnTUtnIf [ xa[r]EscLM cCTUtTOntnfDIul Lcul Vcul [j] Tpeul xa[rul] xa[r]cL gvif CTUca[r]LSsTFOtfDnca[r]Iulc ] jmul[jul] VcTOcul jmul : fDnTOcmu
```

**Warning**: We make utmost efforts to ensure that the compiler works correctly regardless of the sequence used, employing fuzz testing to find any abnormalities, but by its very nature the default sequence is receiving a lot more coverage and problems with custom sequences are much more likely to remain undetected.
For this reason, while the new sequence can also be used with older compilers, we recommend extreme care while doing so.
In particular, the new sequence is susceptible to the [FullInliner Non-Expression-Split Argument Evaluation Order Bug](https://soliditylang.org/blog/2023/07/19/full-inliner-non-expression-split-argument-evaluation-order-bug/), which is not an issue for the recent versions, but would cause problems on versions older than v0.8.21.

> ### Replacement of internal JSON Library

Reviewer 3: ```suggestion
### Replacement of the internal JSON library
```

In this release we also replaced our internal JSON library [`jsoncpp`](https://github.com/open-source-parsers/jsoncpp) with [`nlohmann::json`](https://github.com/nlohmann/json).

Because of that, the formatting of the JSON output slightly changed, where it also became more strict with UTF-8 encodings.
The old [`jsoncpp`](https://github.com/open-source-parsers/jsoncpp) allowed some invalid UTF-8 sequences, but also did not handle them properly.

> However, we don't expect it to create any problems because vast majority of implementation assume UTF-8 anyway.

Reviewer 3: How about this?
```suggestion
However, we don't expect it to create problems in practice because the vast majority of implementations assume UTF-8 anyway.
```
